### ISSUE TITLE
High memory usage?

### ISSUE THREAD

**Issue by jrevels** [denizyuret/AutoGrad.jl#3]
It's great to see Julia's collection of AD tools grow!

I tried to take the gradient of the below function, but it seemed to hang for several minutes before Julia crashes:

```julia
julia> rosenbrock(x) = sum(map((i, j) -> (1 - j)^2 + 100*(i - j^2)^2, x[2:end], x[1:end-1]))
rosenbrock (generic function with 1 method)

julia> g = AutoGrad.grad(rosenbrock)
(::gradfun) (generic function with 1 method)

julia> x = rand(100000);

julia> @time g(x) # just sits here for several minutes, then crashes
```

It seems to be because of the sheer amount of memory being used. I know that reverse-mode, in general, can be a memory hog, but Python's `autograd` library produces this gradient in about 4 seconds on my machine, so I'm suspecting there's a bug somewhere.

**Comment by denizyuret** [comment]
Thanks Jarrett:

The vectorized version works:

rosenbrock2(x) = sum((1-x[1:end-1]).^2 + 100*(x[2:end]-x[1:end-1].^2).^2)

julia> x = rand(100000);

julia> @time rosenbrock2(x) => 0.005783 seconds (29 allocations: 7.630 MB)


Your version creates several recorded primitive operations for each
element, with their input, output, gradients etc.  But this should be a
problem for the Python version as well...  Can you send me the Python
example?

best,

deniz

**Comment by jrevels** [comment]
Ah, the ~4 second Python runtime came from a vectorized version as well, now that I recall. Awesome speed up here versus that version!

Here's a fairer comparison versus my original Julia example:

```python
from itertools import izip
from autograd import grad
import autograd.numpy as np

def rosenbrock(x):
    return sum((1 - j)**2 + 100*(i - j**2)**2 for i, j in izip(x[1:],x[0:-1]))

x = np.random.rand(100000)
g = grad(rosenbrock)
g(x)
```

The above code takes about ~20 seconds to run on my machine, but doesn't run out of memory.

**Comment by denizyuret** [comment]
I tracked down the problem: the gradient for getindex and gradient summations were implemented inefficiently.  I fixed both issues.  The iterative example should work relatively fast now and not blow up the memory.  Thanks for noticing this!

**Comment by jrevels** [comment]
Cool! Seems like there might be a bug on master though? Just pulled master to try it again and got the following error (I'm on Julia v0.5, but I get the same error on v0.4): 

```julia
julia> rosenbrock(x) = sum(map((i, j) -> (1 - j)^2 + 100*(i - j^2)^2, x[2:end], x[1:end-1]))
rosenbrock (generic function with 1 method)

julia> g = AutoGrad.grad(rosenbrock)
(::gradfun) (generic function with 1 method)

julia> x = rand(100000);

julia> @time g(x)
ERROR: MethodError: Cannot `convert` an object of type Float64 to an object of type Void
This may have arisen from a call to the constructor Void(...),
since type constructors fall back to convert methods.
 in setindex!(::Array{Void,1}, ::Array{Float64,1}, ::UnitRange{Int64}) at ./array.jl:433
 in get!(::Array{Void,1}, ::Array{Float64,1}, ::Tuple{UnitRange{Int64}}, ::Void) at ./abstractarray.jl:940
 in get(::Array{Float64,1}, ::Tuple{UnitRange{Int64}}, ::Void) at ./abstractarray.jl:944
 in sum_outgrads(::Array{Float64,1}, ::AutoGrad.OneHot) at /Users/jarrettrevels/.julia/v0.5/AutoGrad/src/interfaces.jl:76
 in backward_pass(::AutoGrad.Value{Array{Float64,1}}, ::AutoGrad.Value{Float64}, ::Array{AutoGrad.Node,1}) at /Users/jarrettrevels/.julia/v0.5/AutoGrad/src/core.jl:195
 in (::AutoGrad.##gradfun#1#3{#rosenbrock,Int64})(::Array{Any,1}, ::Function, ::Array{Float64,1}, ::Vararg{Array{Float64,1},N}) at /Users/jarrettrevels/.julia/v0.5/AutoGrad/src/core.jl:43
 in (::AutoGrad.#gradfun#2)(::Array{Float64,1}, ::Vararg{Array{Float64,1},N}) at /Users/jarrettrevels/.julia/v0.5/AutoGrad/src/core.jl:43
```

**Comment by denizyuret** [comment]
Hah, I ignored the first rule of bug fixing: actually check whether the code in the bug report works!

I had a for-loop version I wrote for the test for some reason:

f3(x)=(s=0; for i=2:length(x); s+=(1-x[i-1])^2 + 100*(x[i]-x[i-1]^2)^2; end; s)

Apparently map handling was not ideal.  Your example works now (I hope) but I need to check things with a fresh brain tomorrow morning to make sure I didn't break anything else.
